concordium-consensus/src/Concordium/KonsensusV1/Consensus/Timeout.hs

```diff
@@ -0,0 +1,560 @@
+        return $ Rejected ObsoleteRound
+    -- If the round or epoch of the qc associated with the timeout message
+    -- is behind the last finalized block then reject the timeout message.
+    -- This happens in case of dishonest behavior as it is checked above that the round of
+    -- the timeout message is greater or equal to the current round, hence an honest
+    -- sender of the timeout message must have progressed their QC to at least the
+    -- last finalized block.
+    | let lastFin = skovData ^. lastFinalized,
+      qcRound tmQuorumCertificate < blockRound lastFin
+        || qcEpoch tmQuorumCertificate < blockEpoch lastFin =
+        return $ Rejected ObsoleteQC
+    -- We check if the epoch that the sender is on is ahead of our current epoch.
+    -- It could be that the QC in the timeout message allows us to advance to the epoch, but
+    -- instead we just trigger catch-up. While this is less efficient in that case, we opt for
+    -- this simpler implementation as it is sufficient and the case is not likely to be common.
+    | tmEpoch > theCurrentEpoch =
+        return CatchupRequired
+    -- Obtain the finalizer information for the signer of the timeout message.
+    | otherwise = withFinalizers $ \finalizationCommittee ->
+        withFinalizer finalizationCommittee $ \case
+            finInfo@FinalizerInfo{..}
+                -- Check whether the signature is ok or not.
+                | not (checkTimeoutMessageSignature finalizerSignKey genesisBlockHash tm) ->
+                    return $ Rejected InvalidSignature
+                -- Consensus runner is not caught up to the round that the timeout message
+                -- refers to. So catch up is required.
+                | tmRound > currentRound,
+                  qcRound tmQuorumCertificate < tmRound - 1 ->
+                    return CatchupRequired
+                | otherwise ->
+                    checkForDuplicate $
+                        getRecentBlockStatus (qcBlock tmQuorumCertificate) skovData >>= \case
+                            -- In this case, the quorum certificate cannot be valid, because otherwise
+                            -- it would already have been rejected for an obsolete QC.
+                            OldFinalized ->
+                                return $ Rejected ObsoleteQCPointer
+                            -- The QC pointer points to a block that has been marked dead.
+                            -- (Dead blocks are either invalid or branch from a pruned block.)
+                            -- Here also the QC cannot be valid.
+                            RecentBlock BlockDead ->
+                                return $ Rejected DeadQCPointer
+                            -- The timeout message is now verified and ready for being retransmitted.
+                            RecentBlock (BlockAliveOrFinalized qcBlock) ->
+                                received finInfo finalizationCommittee (Present qcBlock)
```

Comment:
In this branch we don't check at any point if we have a record for this QC in `roundexistingQCs` (the first thing we do in `executeTimeoutMessage` when `Present qcBlock` is supplied as an argument is to run `checkQC`. So the result is that we will check systematically check all QCs in timeouts, even when they are all pointing to the same block, which is the expected case.

I'm putting the comment in this function and not in `executeTimeoutMessage`, because I think we should fix it here. Since both branches (`RecentBlock BlockPendingOrUnknown` and `RecentBlock (BlockAliveOrFinalized qcBlock)`) require us to do exactly the same check, my suggestion is to first to do this check before even calling `getRecentBlockStatus`, like that the same code only appears once and we also avoid some unnecessary calls to `getRecentBlockStatus`.

1. `checkForDuplicate`
2. `qcRound tmQuorumCertificate <= skovData ^. roundStatus . rsHighestCertifiedBlock . to cbRound` AND `Just (QuorumCertificateCheckedWitness certEpoch) <- skovData ^. roundExistingQuorumCertificate (qcRound tmQuorumCertificate)`
3. If test in 2 fails, `getRecentBlockStatus`
4. If `getRecentBlockStatus` results in `RecentBlock BlockPendingOrUnknown`, we do catch-up
5. If `getRecentBlockStatus` results in `RecentBlock (BlockAliveOrFinalized qcBlock)`, we call `received finInfo finalizationCommittee (Present qcBlock)`
6. If `getRecentBlockStatus` results in `OldFinalized` or `RecentBlock BlockDead`, we reject
7. If test in 2 succeeds, we check if `qcEpoch tmQuorumCertificate == certEpoch`
8. If test in 7 succeeds, we call `received finInfo finalizationCommittee Absent`
9. If test in 7 fails, we reject.

Reply:
Fixed in https://github.com/Concordium/concordium-node/pull/865